cluster/images/etcd/migrate/migrate.go

```diff
@@ -46,94 +44,42 @@ a target etcd version, this tool will upgrade or downgrade the etcd data from th
 	opts = migrateOpts{}
 )
 
-type migrateOpts struct {
-	name                 string
-	port                 uint64
-	peerListenUrls       string
-	peerAdvertiseUrls    string
-	binDir               string
-	dataDir              string
-	bundledVersionString string
-	etcdDataPrefix       string
-	ttlKeysDirectory     string
-	initialCluster       string
-	targetVersion        string
-	targetStorage        string
-	etcdServerArgs       string
-}
-
 func main() {
-	flags := migrateCmd.Flags()
-	flags.StringVar(&opts.name, "name", "", "etcd cluster member name. Defaults to etcd-{hostname}")
-	flags.Uint64Var(&opts.port, "port", defaultPort, "etcd client port to use during migration operations. This should be a different port than typically used by etcd to avoid clients accidentally connecting during upgrade/downgrade operations.")
-	flags.StringVar(&opts.peerListenUrls, "listen-peer-urls", "", "etcd --listen-peer-urls flag, required for HA clusters")
-	flags.StringVar(&opts.peerAdvertiseUrls, "initial-advertise-peer-urls", "", "etcd --initial-advertise-peer-urls flag, required for HA clusters")
-	flags.StringVar(&opts.binDir, "bin-dir", "/usr/local/bin", "directory of etcd and etcdctl binaries, must contain etcd-<version> and etcdctl-<version> for each version listed in bindled-versions")
-	flags.StringVar(&opts.dataDir, "data-dir", "", "etcd data directory of etcd server to migrate")
-	flags.StringVar(&opts.bundledVersionString, "bundled-versions", "", "comma separated list of etcd binary versions present under the bin-dir")
-	flags.StringVar(&opts.etcdDataPrefix, "etcd-data-prefix", "/registry", "etcd key prefix under which all objects are kept")
-	flags.StringVar(&opts.ttlKeysDirectory, "ttl-keys-directory", "", "etcd key prefix under which all keys with TTLs are kept. Defaults to {etcd-data-prefix}/events")
-	flags.StringVar(&opts.initialCluster, "initial-cluster", "", "comma separated list of name=endpoint pairs. Defaults to etcd-{hostname}=http://localhost:2380")
-	flags.StringVar(&opts.targetVersion, "target-version", "", "version of etcd to migrate to. Format must be '<major>.<minor>.<patch>'")
-	flags.StringVar(&opts.targetStorage, "target-storage", "", "storage version of etcd to migrate to, one of: etcd2, etcd3")
-	flags.StringVar(&opts.etcdServerArgs, "etcd-server-extra-args", "", "additional etcd server args for starting etcd servers during migration steps, --peer-* TLS cert flags should be added for etcd clusters with more than 1 member that use mutual TLS for peer communication.")
+	parseFlags(migrateCmd.Flags(), &opts)
+	if err := opts.validateAndDefault(); err != nil {
```

Comment:
I did a quick search. It looks like the flags are parsed during calling `migrateCmd.Execute()`. If we set a flag via command line, would it appear unset inside our custom-defined `validateAndDefault()`?

https://github.com/spf13/cobra/blob/04318720db1743b8488c86b2f7dca6d9663cb2f2/command.go#L769

Reply:
Good catch  (I focused on testing with env vars since the migrate-if-needed.sh no longer sets flags, but we shouldn't break it.

Should be fixed now.